t/data_migration_pipeline/main.go

```diff
@@ -52,18 +57,26 @@ func parseDataModel(record []string) *DataModel {
 }
 
 // emitResult emits data models to be written to Spanner
-func emitResult(s beam.Scope, lines beam.PCollection) beam.PCollection {
+func emitResult(ctx context.Context, s beam.Scope, lines beam.PCollection) beam.PCollection {
 	dataModels := beam.ParDo(s, func(line string, emit func(*DataModel)) {
 		reader := csv.NewReader(strings.NewReader(line))
 		csvLine, err := reader.Read()
 		if err != nil {
 			log.Fatalf("Failed to read record: %v", err)
 		}
 		emit(parseDataModel(csvLine))
+		count.Inc(ctx, 1)
 	}, lines)
+
 	return dataModels
 }
 
+func ctxWithPtransformID(id string) context.Context {
+	ctx := context.Background()
+	ctx = metrics.SetPTransformID(ctx, id)
+	return ctx
+}
```

Comment:
This seems incorrect. We should be threading through a parent context (e.g. from `main` or `realMain`) to properly handle cancellation.

Reply:
For example, I'd expect:

```go
func main() {
  ctx, cancel := signal.NotifyContext(context.Background(),
    syscall.SIGINT, syscall.SIGTERM)
  defer cancel()

  ctx = metrics.SetPTransformID(ctx, "pipeline")

  // everything that accepts a `context.Context` uses `ctx`...

  direct.Execute(ctx, p)
}
```

Reply:
Updated